// rollup-plugin-stub/dist/rollup-plugin-stub.mjs
import MagicString from 'magic-string';
import { createFilter } from 'rollup-pluginutils';
import { parse } from 'acorn';

function stub() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var filter = createFilter(options.include, options.exclude);

  return {
    transform: function transform(code, id) {
      if (!filter(id)) return null;

      var ast = void 0;

      try {
        ast = parse(code, {
          ecmaVersion: 6,
          sourceType: 'module'
        });
      } catch (err) {
        err.message += ' in ' + id;
        throw err;
      }

      var sourceMap = options.sourceMap !== false;
      var magicString = new MagicString(code);
      var bindings = [];

      ast.body.forEach(function (node) {
        if (node.type === 'ExportNamedDeclaration') {
          // Handle null declarations
          // Example: `export { varName as default };`
          if (node.declaration == null) {
            return;
          }
          // Collect binding names and ensure we can reassign them.
          if (node.declaration.type === 'VariableDeclaration') {
            if (node.declaration.kind === 'const') {
              magicString.overwrite(node.declaration.start, node.declaration.start + node.declaration.kind.length, 'var');
            }
            node.declaration.declarations.forEach(function (declaration) {
              bindings.push(declaration.id.name);
            });
          } else if (node.declaration.type === 'FunctionDeclaration') {
            bindings.push(node.declaration.id.name);
          } else if (node.declaration.type === 'ClassDeclaration') {
            bindings.push(node.declaration.id.name);
          }
        }
      });

      bindings.forEach(function (binding) {
        magicString.append('\nvar __original_' + binding + ', ' + binding + '_stubbed;' + ('\nexport function stub_' + binding + '(__stubValue__) { if (!' + binding + '_stubbed) { __original_' + binding + ' = ' + binding + '; ' + binding + '_stubbed = true; } ' + binding + ' = __stubValue__; }') + ('\nexport function reset_' + binding + '() { if (' + binding + '_stubbed) { ' + binding + '_stubbed = false; ' + binding + ' = __original_' + binding + '; } }'));
      });

      code = magicString.toString();
      var map = sourceMap ? magicString.generateMap() : null;

      return { code: code, map: map };
    }
  };
}

export default stub;
